bentoml/_internal/io_descriptors/numpy.py

```diff
@@ -251,6 +251,254 @@ async def to_http_response(self, obj: ext.NpNDArray, ctx: Context | None = None)
+        res = self._verify_ndarray(res, BadInput)
+        return res
+
+    def is_supported(self, data):
+        """
+        Checks if the given type is within `supported_datatypes` dictionary
+        """
+        import datetime
+
+        import numpy as np
+
+        supported_datatypes = {
+            np.int32: "sint32_",
+            np.int64: "sint64_",
+            np.uint32: "uint32_",
+            np.uint64: "uint64_",
+            np.float32: "float_",
+            np.float64: "double_",
+            np.bool_: "bool_",
+            np.bytes_: "bytes_",
+            np.str_: "string_",
+            np.ndarray: "array_",
+            np.datetime64: "timestamp_",
+            np.timedelta64: "duration_"
+            # TODO : complex types, lower byte integers(8,16)
+        }
+
+        found_dtype = ""
+        for key in supported_datatypes:
+            if np.dtype(type(data)) == key:
+                found_dtype = supported_datatypes[key]
+                if found_dtype == "array_":
+                    if isinstance(data, datetime.datetime) or isinstance(
+                        data, datetime.date
+                    ):
+                        found_dtype = "timestamp_"
+                    elif isinstance(data, datetime.timedelta):
+                        found_dtype = "duration_"
+                break
+
+        return found_dtype
+
+    def create_tuple_proto(self, tuple):
+        """
+        Convert given tuple list or tuple array to protobuf
+        """
+        import datetime
+
+        import numpy as np
+        from google.protobuf.duration_pb2 import Duration
+        from google.protobuf.timestamp_pb2 import Timestamp
+
+        from ..generated_pb import io_descriptors_pb2
+
+        if len(tuple) == 0:
+            raise ValueError("Provided tuple is either empty or invalid.")
+
+        tuple_arr = []
+        for item in tuple:
+            dtype = self.is_supported(item)
+
+            if not dtype:
+                raise ValueError(
+                    f'Invalid datatype "{type(item).__name__}" within tuple.'
+                )
+            elif dtype == "timestamp_":
+                if isinstance(item, np.datetime64):
+                    item = item.astype(datetime.datetime)
+                if isinstance(item, datetime.date):
+                    item = datetime.datetime(item.year, item.month, item.day)
+                t = Timestamp()
+                t.FromDatetime(item)
+                tuple_arr.append(io_descriptors_pb2.Value(**{"timestamp_": t}))
+            elif dtype == "duration_":
+                if isinstance(item, np.timedelta64):
+                    item = item.astype(datetime.timedelta)
+                d = Duration()
+                d.FromTimedelta(item)
+                tuple_arr.append(io_descriptors_pb2.Value(**{"duration_": d}))
+            elif dtype == "array_":
+                if not all(isinstance(x, type(item[0])) for x in item):
```

Comment:
This will throw IndexError if item is not a iterable, and also if item is a string then this would fail?

Reply:
Line 414 is only reachable if the data has `dtype` "array_". If the `dtype` is "array_" it should be iterable.